src/btree/bt_read.c

```diff
@@ -84,6 +84,119 @@ __evict_force_check(WT_SESSION_IMPL *session, WT_REF *ref)
     return (__wt_page_can_evict(session, ref, NULL));
 }
 
+/*
+ * __bt_reconstruct_delta --
+ *     Reconstruct delta on a page
+ */
+static int
+__bt_reconstruct_delta(WT_SESSION_IMPL *session, WT_REF *ref, WT_ITEM *delta)
+{
+    WT_CELL_UNPACK_DELTA unpack;
+    WT_CURSOR_BTREE cbt;
+    WT_DECL_RET;
+    WT_DELTA_HEADER *header;
+    WT_ITEM key, value;
+    WT_PAGE *page;
+    WT_ROW *rip;
+    WT_UPDATE *first_upd, *upd, *standard_value, *tombstone;
+    size_t size, tmp_size, total_size;
+
+    header = (WT_DELTA_HEADER *)delta->data;
+    total_size = 0;
+    page = ref->page;
+
+    __wt_btcur_init(session, &cbt);
+    __wt_btcur_open(&cbt);
+
+    WT_CELL_FOREACH_DELTA(session, header, unpack)
+    {
+        key.data = unpack.key;
+        key.size = unpack.key_size;
+        upd = standard_value = tombstone = NULL;
+        size = 0;
+
+        /* Search the page and apply the modification. */
+        WT_ERR(__wt_row_search(&cbt, &key, true, ref, true, NULL));
+        /*
+         * We apply deltas from newest to oldest, ignore keys that have already got a delta update.
+         */
+        if (cbt.compare == 0) {
+            if (cbt.ins != NULL) {
+                if (cbt.ins->upd != NULL && F_ISSET(cbt.ins->upd, WT_UPDATE_RESTORED_FROM_DELTA))
+                    continue;
+            } else {
+                rip = &page->pg_row[cbt.slot];
+                first_upd = WT_ROW_UPDATE(page, rip);
+                if (first_upd != NULL && F_ISSET(first_upd, WT_UPDATE_RESTORED_FROM_DELTA))
+                    continue;
+            }
+        }
+
+        if (F_ISSET(&unpack, WT_DELTA_IS_DELETE)) {
+            WT_RET(__wt_upd_alloc_tombstone(session, &tombstone, &tmp_size));
+            F_SET(tombstone, WT_UPDATE_DURABLE | WT_UPDATE_RESTORED_FROM_DELTA);
+            size += tmp_size;
+            upd = tombstone;
+        } else {
+            value.data = unpack.value;
+            value.size = unpack.value_size;
+            WT_RET(__wt_upd_alloc(session, &value, WT_UPDATE_STANDARD, &standard_value, &tmp_size));
+            standard_value->start_ts = unpack.tw.start_ts;
+            standard_value->durable_ts = unpack.tw.durable_start_ts;
+            F_SET(standard_value, WT_UPDATE_DURABLE | WT_UPDATE_RESTORED_FROM_DELTA);
+            size += tmp_size;
+
+            if (WT_TIME_WINDOW_HAS_STOP(&unpack.tw)) {
+                WT_RET(__wt_upd_alloc_tombstone(session, &tombstone, &tmp_size));
+                tombstone->start_ts = unpack.tw.stop_ts;
+                tombstone->durable_ts = unpack.tw.durable_stop_ts;
+                F_SET(tombstone, WT_UPDATE_DURABLE | WT_UPDATE_RESTORED_FROM_DELTA);
+                tombstone->next = standard_value;
+                upd = tombstone;
```

Comment:
I don't know the btree code well, but it seems like this if statement could be moved up into the else above, like:
```
   } else if (WT_TIME_WINDOW_HAS_STOP()) {
```
In its current position, we're doing some extra work above this that looks like it won't get used.  Related, I think `size += tmp_size` above these lines will prove to be inaccurate.

Reply:
It is not extra work. In this part, we either instantiate a full update or a tombstone -> a full update depending on whether we have a stop timestamp.